ccl/sqlccl: TestBackupRestoreControlJob failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/4fd3e09aaa7974a7fb6f81853d717c25b082d878

Parameters:
```
TAGS=deadlock
GOFLAGS=
```

Stress build found a failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=455324&tab=buildLog

```
I171223 10:12:01.624626 9732186 sql/event_log.go:113  [client=127.0.0.1:38614,user=root,n1] Event: "create_database", target: 61, info: {DatabaseName:cancelimport Statement:CREATE DATABASE cancelimport User:root}
I171223 10:12:01.660555 9451500 storage/replica_command.go:1231  [split,n3,s3,r117/2:/{Table/60/1/9…-Max}] initiating a split of this range at key /Table/61 [r118]
I171223 10:12:01.772297 9451388 storage/replica_proposal.go:195  [n3,s3,r117/2:/{Table/60/1/9…-Max}] new range lease repl=(n3,s3):2 start=1514023919.845143483,0 epo=1 pro=1514023919.845212063,0 following repl=(n3,s3):2 start=1514023919.845143483,0 epo=1 pro=1514023919.845212063,0
W171223 10:12:01.837455 10335420 storage/engine/mvcc.go:1871  [n1,s1,r12/1:/Table/1{5-6}] unable to find value for /Table/15/2/"pending"/2017-12-23T10:12:01.69962Z/307866899888472065/0 @ 1514023921.774163341,1
W171223 10:12:01.956656 10338084 storage/engine/mvcc.go:1900  [n1,s1,r12/1:/Table/1{5-6}] unable to find value for /Table/15/2/"pending"/2017-12-23T10:12:01.774163Z/307866899888472065/0 @ 1514023921.888943849,1
I171223 10:12:01.959654 10340270 ccl/sqlccl/csv.go:420  [client=127.0.0.1:38544,user=root,n1] could not fetch file size; falling back to per-file progress: bad ContentLength: -1
W171223 10:12:02.124131 10344974 storage/engine/mvcc.go:1900  [n1,s1,r12/1:/Table/1{5-6}] unable to find value for /Table/15/2/"running"/2017-12-23T10:12:01.774163Z/307866899888472065/0 @ 1514023922.083762405,0
W171223 10:12:02.313412 9385839 storage/engine/mvcc.go:1900  [n1,s1,r12/1:/Table/1{5-6}] unable to find value for /Table/15/2/"pending"/2017-12-23T10:12:02.164567Z/307866901428699137/0 @ 1514023922.242377124,1
I171223 10:12:02.314676 10353592 ccl/sqlccl/csv.go:420  [client=127.0.0.1:38614,user=root,n1] could not fetch file size; falling back to per-file progress: bad ContentLength: -1
I171223 10:12:02.367237 9385509 server/status/runtime.go:219  [n1] runtime stats: 485 MiB RSS, 564 goroutines, 47 MiB/225 MiB/339 MiB GO alloc/idle/total, 49 MiB/75 MiB CGO alloc/total, 6877.25cgo/sec, 1.10/0.05 %(u/s)time, 0.03 %gc (22x)
I171223 10:12:03.367403 10391187 storage/replica_command.go:1231  [n3,s3,r118/2:/{Table/61-Max}] initiating a split of this range at key /Table/62 [r119]
I171223 10:12:03.408160 9451374 storage/replica_proposal.go:195  [n3,s3,r118/2:/{Table/61-Max}] new range lease repl=(n3,s3):2 start=1514023919.845143483,0 epo=1 pro=1514023919.845212063,0 following repl=(n3,s3):2 start=1514023919.845143483,0 epo=1 pro=1514023919.845212063,0
I171223 10:12:03.456670 10397143 storage/replica_command.go:1231  [n3,s3,r119/2:/{Table/62-Max}] initiating a split of this range at key /Table/62/1/307866901922906112/0 [r120]
I171223 10:12:03.576131 9451339 storage/replica_proposal.go:195  [n3,s3,r119/2:/{Table/62-Max}] new range lease repl=(n3,s3):2 start=1514023919.845143483,0 epo=1 pro=1514023919.845212063,0 following repl=(n3,s3):2 start=1514023919.845143483,0 epo=1 pro=1514023919.845212063,0
I171223 10:12:03.621916 10405811 storage/replica_command.go:1231  [n3,s3,r120/2:/{Table/62/1/3…-Max}] initiating a split of this range at key /Table/62/1/307866903980539904/0/NULL [r121]
I171223 10:12:03.663422 9451394 storage/replica_proposal.go:195  [n3,s3,r120/2:/{Table/62/1/3…-Max}] new range lease repl=(n3,s3):2 start=1514023919.845143483,0 epo=1 pro=1514023919.845212063,0 following repl=(n3,s3):2 start=1514023919.845143483,0 epo=1 pro=1514023919.845212063,0
I171223 10:12:03.827208 10417743 util/stop/stopper.go:474  quiescing; tasks left:
1      storage.intentResolver: processing intents
W171223 10:12:03.827656 10417689 storage/replica.go:2795  [n1,s1,r12/1:/Table/1{5-6}] shutdown cancellation after 0.0s of attempting command ResolveIntent [/Table/15/1/307866905046024193/0,/Min), ResolveIntent [/Table/15/2/"running"/2017-12-23T10:12:03.268963Z/307866905046024193/0,/Min), ResolveIntent [/Table/15/2/"succeeded"/2017-12-23T10:12:03.268963Z/307866905046024193/0,/Min)
W171223 10:12:03.829147 9457417 storage/raft_transport.go:461  [n3] raft transport stream to node 1 failed: rpc error: code = FailedPrecondition desc = grpc: the client connection is closing
W171223 10:12:03.829698 10417675 storage/intent_resolver.go:330  [n3,s3] failed to resolve intents: failed to send RPC: sending to all 3 replicas failed; last error: {<nil> rpc error: code = FailedPrecondition desc = grpc: the client connection is closing}
W171223 10:12:03.829865 9466313 storage/raft_transport.go:461  [n3] raft transport stream to node 2 failed: rpc error: code = FailedPrecondition desc = grpc: the client connection is closing
W171223 10:12:03.830269 9454795 storage/raft_transport.go:461  [n2] raft transport stream to node 1 failed: rpc error: code = Unavailable desc = transport is closing
W171223 10:12:03.830417 9465941 storage/raft_transport.go:461  [n2] raft transport stream to node 3 failed: rpc error: code = FailedPrecondition desc = grpc: the client connection is closing
W171223 10:12:03.830941 9454740 storage/raft_transport.go:461  [n1] raft transport stream to node 2 failed: EOF
W171223 10:12:03.831072 9457373 storage/raft_transport.go:461  [n1] raft transport stream to node 3 failed: EOF
I171223 10:12:03.832209 10417741 util/stop/stopper.go:474  quiescing; tasks left:
1      node.Node: batch
E171223 10:12:03.836100 9447333 sql/jobs/registry.go:208  error while adopting jobs: node unavailable; try another peer
E171223 10:12:03.843427 9451788 sql/jobs/registry.go:208  error while adopting jobs: node unavailable; try another peer
```

--- a/pkg/internal/client/txn.go
+++ b/pkg/internal/client/txn.go
@@ -773,20 +773,22 @@ func (txn *Txn) Exec(
 			}
 		}
 
-		if _, ok := err.(*roachpb.UnhandledRetryableError); ok {
+		cause := errors.Cause(err)
+
+		if _, ok := cause.(*roachpb.UnhandledRetryableError); ok {
 			// We sent transactional requests, so the TxnCoordSender was supposed to
 			// turn retryable errors into HandledRetryableTxnError.
 			log.Fatalf(ctx, "unexpected UnhandledRetryableError at the txn.Exec level: %s", err)
 		}
 
-		retErr, retryable := err.(*roachpb.HandledRetryableTxnError)
+		retErr, retryable := cause.(*roachpb.HandledRetryableTxnError)
 		if retryable && !txn.IsRetryableErrMeantForTxn(*retErr) {
 			// Make sure the txn record that err carries is for this txn.
 			// If it's not, we terminate the "retryable" character of the error. We
 			// might get a HandledRetryableTxnError if the closure ran another
 			// transaction internally and let the error propagate upwards.
 			return errors.Wrapf(
-				retErr,
+				err,
 				"retryable error from another txn. Current txn ID: %v", txn.Proto().ID)
 		}
 		if !opt.AutoRetry || !retryable {

